pkg/trigram/trigram.go

```diff
@@ -0,0 +1,103 @@
+package trigram
+
+import (
+	"golang.org/x/text/unicode/norm"
+	"math/rand"
+	"strings"
+	"time"
+	"unicode"
+	"unicode/utf8"
+)
+
+type Trigram struct {
+	Word   string
+	Offset uint64
+}
+
+const N = 3
+
+// BuildTrigram makes two passes
+//
+//	1 - splits by white space and keeps track of the positions
+//	2 - performs a sliding window and builds trigrams
+
+func normalizeToAscii(s string) (string, map[int]int) {
+	ogOffsets := make(map[int]int)
+
+	var b strings.Builder
+	norm := norm.NFKD.String(s)
+
+	additionalOffsets := 0
+
+	newIndex := 0
+
+	for i, r := range norm {
+		if utf8.RuneLen(r) > 1 {
+			additionalOffsets += utf8.RuneLen(r) - 1
+		}
+
+		if r <= unicode.MaxASCII {
+			b.WriteRune(r)
+			ogOffsets[newIndex] = i - additionalOffsets
+			newIndex++
+		}
+
+	}
+	return b.String(), ogOffsets
+}
+
+func Shuffle(trigrams []Trigram) []Trigram {
+	soup := make([]Trigram, len(trigrams))
+	copy(soup, trigrams)
+
+	rand.Seed(time.Now().UnixNano())
```

Comment:
Can we seed the randomness based on the hash of a string? https://stackoverflow.com/a/13582881/86433

This way, the same string will always shuffle to the same result making it deterministically random.

Reply:
Sure but more importantly, I assume this is what we want for the typescript version